runtime/runtime.go

```diff
@@ -3,11 +3,23 @@ package runtime
 import (
 	"fmt"
 	"google.golang.org/protobuf/proto"
+	"google.golang.org/protobuf/reflect/protoreflect"
 	"google.golang.org/protobuf/runtime/protoiface"
 	"io"
 	"math/bits"
 )
 
+type CustomType interface {
+	UnmarshalBytes(in protoiface.UnmarshalInput, b []byte) (out protoiface.UnmarshalOutput, err error)
+	MarshalBytes(in protoiface.MarshalOutput) (out protoiface.MarshalOutput, err error)
+	Size(in protoiface.SizeInput) (out protoiface.SizeOutput)
+	Set(value protoreflect.Value)
+	Get() protoreflect.Value
+	Clear()
+	IsSet() bool
+	Mutable() protoreflect.Value
+}
```

Comment:
Not sure we need to make it this complex. On one hand this allows caching of the `protoreflect.Value` + custom type, but feels like it mixes too many concerns. Can you say a little bit more about the use case for messages and `Mutable`?

My thinking was that we _only_ handle `string` or `bytes` types for now because those are the only cases we have with a very simple two method Unmarshal/Marshal interface, and that a scalar can only work with one proto type (string or bytes)

Reply:
We can see it in this way:
Do we want a custom type to be capable of overriding a message (ex we want to override anypb.Any with pulsar.Any) ?
If this is the intention then this is the only API which does not require multiple implementations of custom type logic that can serve all purposes.
If not then we can simplify by removing unmarshal/marshal/size inputs and Mutable

Reply:
So as we discussed I don't think we want to extend this for full Message type overriding (for Any's we can just replace the type name).

For simple scalars, we could use something like I described in #2 for `CustomStringType` or `CustomBytesType` or even something as simple as:

```go
type CustomType interface {
  Set(value protoreflect.Value)
  Get() protoreflect.Value
}
```

Reply:
doing one custom type for each primitive is going to be a big chunk of work codegen wise, this is why I was going for a more generalized interface. Plus custom types are going to be rare and I think they're an advanced level.

What if we simplified the current custom type interface to this:

```go
type CustomType interface {
	UnmarshalBytes(b []byte) error
	MarshalBytes() ([]byte, error)
	Size() int
	Set(value protoreflect.Value)
	Get() protoreflect.Value
	Clear()
	IsSet() bool
}
```

Reply:
which approach from #2 are you thinking of going with? why do we need both Unmarshal/Marshal + Set/Get? I'm not clear as to why we couldn't deal everything with Set/Get? With Unmarshal/Marshal the type doesn't know if it was passed a string, bytes or int64 say.

Reply:
Also do we need IsSet and Clear? For the Int and Dec case an empty string should we 0 and not NaN right? Maybe for addresses though IsSet has a meaning...

Reply:
Those methods are needed for Protoreflect.Clear and Has

Reply:
right, but if we go with the second approach in #2 we don't need those.

I guess I'm thinking with whichever approach we take we do want the underlying proto value to get cached either by the proto message itself or the custom type

Reply:
using the 1st approach is simpler to implement (from codegen perspective).

second approach would need more time because generating that logic is a lot more complex, and needs to be handled with a case by case basis in every method.

Also I think both obtain the same result, with the custom type implementation being a little harder to implement - but I think that's fine, I don't expect every dev to want to implement its own custom type.

Reply:
Also another reason I'd go with 1) is because in the sdk that's how we deal with custom types currently.  I think the API is more dev friendly and less breaking.

Reply:
do you think it's reasonable to expect custom types to cache the `protoreflect.Value`?

> second approach would need more time because generating that logic is a lot more complex, and needs to be handled with a case by case basis in every method.

why is this the case? it just adds two fields plus a getter/setter pair without impacting any of the other marshaling/protoreflect code